### PR TITLE
Use hydra for authorization flow

### LINKED ISSUE #1180: Don't store username and password hash in the client.prefs file
The client stores the hash of the password in the `client.prefs` text file.
The file stores configuration related info mainly and may be requested for troubleshooting which causes the user unwillingly share his/her password.

I think it would be better to store the password hash in a separate file, in a gziped binary file even along with the username for better safety when `Automatic login` is requested.

### LINKED ISSUE #1533: Authenticate with lobby server using API OAuth tokens
[Issuehunt badges](https://issuehunt.io/r/FAForever/downlords-faf-client/issues/1533)

See the related PR https://github.com/FAForever/server/pull/473.

The new command replaces `hello` with a command that works almost the same, except instead of taking a username and password, it takes an OAuth token signed by the API. In order to solve the problem of what to do with the IRC password, we thought of having the server generate a password at random, and send it back to the client.

New auth message:
```json
{
    "command": "auth",
    "token": "<token_string>",
    "unique_id": "<unique_id_string>"
}
```

IRC password response:
```json
{
    "command": "irc_password",
    "password": "<new_irc_password>"
}
```
Some details will still need to be worked out.

**Wanna have the bug fixed quickly?**
Visit Issue hunt...
[Issue hunt](https://issuehunt.io/r/FAForever/downlords-faf-client)

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $5.00)

- [katharsas katharsas](https://issuehunt.io/u/katharsas) ($5.00)

### Submitted pull Requests
- [#2175 Use hydra for authorization flow](https://issuehunt.io/r/FAForever/downlords-faf-client/pull/2175)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/FAForever/downlords-faf-client/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by Sheikah45**
This is the start of using the hydra service for login
Fixes #1180
Fixes #1533

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#1533: Authenticate with lobby server using API OAuth tokens](https://issuehunt.io/repos/32919066/issues/1533)
---
</details>

### CODE REVIEW

**Comment by Sheikah45** [comment]
Works with https://github.com/FAForever/server/pull/784

File: src/main/java/com/faforever/client/api/TokenService.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  public TokenService(PreferencesService preferencesService) {
+    this.preferencesService = preferencesService;
+
+    restTemplate = new RestTemplateBuilder().
+        build();
+  }
+
+  @SneakyThrows
+  public OAuth2AccessToken getRefreshedToken() {
+    if (tokenCache == null || tokenCache.isExpired()) {
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/api/TokenService.java]: Did you think about the case where the refreshtoken expired and you gotta log out the user 🤔

Comment (1-alex98) [reply src/main/java/com/faforever/client/api/TokenService.java]: Hmm would also be really rude tho logging him out during a game where being logged in is needed to communicate with the server. But I was once told Brutus plans on making it 30 days so frequent players will definitely experience that case. 🤔

File: src/main/java/com/faforever/client/login/LoginController.java [hunk 2]
```diff
@@ -183,35 +157,85 @@ public Endpoints fromString(String string) {
...
+            if (refreshToken != null) {
+              userService.loginWithRefreshToken(refreshToken);
+            }
           });
     } else {
-      loginAllowed = true;
       initializeFuture = CompletableFuture.completedFuture(null);
     }
+
+    webViewConfigurer.configureWebView(loginWebView);
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/login/LoginController.java:176]: Would maybe refactor this method into several

File: src/main/java/com/faforever/client/api/TokenService.java (hunk 1 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/api/TokenService.java]: I went the route of logging people out of the api and requesting them to reenter their credentials. It does not log you out of the server unless you relogin as a different player. This should prevent people from losing connection in the middle of games.

**Comment by Sheikah45** [comment]
Alright this should be good to go now outside of tests. So gimme reviews @1-alex98  @Brutus5000

**Pull request merged**